File: research/verdict_research/features/reviewer_graph.py
from dataclasses import dataclass


@dataclass(frozen=True)
class ReviewForReviewerGraph:
    reviewer_id: str | None


@dataclass
class ReviewerGraphResult:
    flagged_review_share: float | None
    flagged_review_count: int
    flagged_reviewer_count: int
    known_reviewer_count: int
    identified_review_count: int


def reviewer_graph_share(
    reviews: list[ReviewForReviewerGraph],
    flagged_reviewer_ids: set[str],
) -> ReviewerGraphResult:
    known: set[str] = set()
    flagged_reviewers: set[str] = set()
    identified_review_count = 0
    flagged_review_count = 0

    for review in reviews:
        reviewer_id = review.reviewer_id
        if reviewer_id is None:
            continue
        identified_review_count += 1
        known.add(reviewer_id)
        if reviewer_id in flagged_reviewer_ids:
            flagged_review_count += 1
            flagged_reviewers.add(reviewer_id)

    return ReviewerGraphResult(
        flagged_review_share=(
            None if identified_review_count == 0 else flagged_review_count / identified_review_count
        ),
        flagged_review_count=flagged_review_count,
        flagged_reviewer_count=len(flagged_reviewers),
        known_reviewer_count=len(known),
        identified_review_count=identified_review_count,
    )

File: extension/src/score/reviewerGraph.ts
export interface ReviewForReviewerGraph {
  reviewerId: string | null;
}

export interface ReviewerGraphResult {
  flaggedReviewShare: number | null;
  flaggedReviewCount: number;
  flaggedReviewerCount: number;
  knownReviewerCount: number;
  identifiedReviewCount: number;
}

export function reviewerGraphShare(
  reviews: readonly ReviewForReviewerGraph[],
  flaggedReviewerIds: ReadonlySet<string>,
): ReviewerGraphResult {
  const known = new Set<string>();
  const flaggedReviewers = new Set<string>();
  let identifiedReviewCount = 0;
  let flaggedReviewCount = 0;

  for (const review of reviews) {
    const id = review.reviewerId;
    if (id === null) {
      continue;
    }
    identifiedReviewCount++;
    known.add(id);
    if (flaggedReviewerIds.has(id)) {
      flaggedReviewCount++;
      flaggedReviewers.add(id);
    }
  }

  return {
    flaggedReviewShare: identifiedReviewCount === 0 ? null : flaggedReviewCount / identifiedReviewCount,
    flaggedReviewCount,
    flaggedReviewerCount: flaggedReviewers.size,
    knownReviewerCount: known.size,
    identifiedReviewCount,
  };
}
